Source/ocdm/open_cdm_impl.cpp

```diff
@@ -0,0 +1,130 @@
+        SessionPrivate(const SessionPrivate&) = delete;
+        SessionPrivate& operator= (const SessionPrivate&) = delete;
+
+        SessionPrivate()
+            : _constructSessionPvt(nullptr)
+            , _destructSessionPvt(nullptr) {
+                Load();
+        }
+        ~SessionPrivate() = default;
+
+        OpenCDMError Construct(struct OpenCDMSession* session, void* &pvtData)
+        {
+            OpenCDMError result = OpenCDMError::ERROR_METHOD_NOT_IMPLEMENTED;
+            if(_constructSessionPvt != nullptr) {
+                result = ( _constructSessionPvt(session, pvtData) == 0 ? OpenCDMError::ERROR_NONE : OpenCDMError::ERROR_UNKNOWN);
+            }
+            return result;
+        }
+
+        OpenCDMError Destruct(struct OpenCDMSession* session, void* &pvtData)
+        {
+            OpenCDMError result = OpenCDMError::ERROR_METHOD_NOT_IMPLEMENTED;
+            if(_destructSessionPvt != nullptr) {
+                result = ( _destructSessionPvt(session, pvtData) == 0 ? OpenCDMError::ERROR_NONE : OpenCDMError::ERROR_UNKNOWN);
+            }
+            return result;
+        }
+
+    private:
+        void Load() {
+            const TCHAR *name = nullptr;
+            Core::Library library(name);
+            if (library.IsLoaded() == true) {
+                _constructSessionPvt = reinterpret_cast<ConstructSessionPrivate>(library.LoadFunction(_T("opencdm_construct_session_private")));
+                _destructSessionPvt = reinterpret_cast<DestructSessionPrivate>(library.LoadFunction(_T("opencdm_destruct_session_private")));
+            }
+
+        }
+
+    private:
+        ConstructSessionPrivate _constructSessionPvt;
+        DestructSessionPrivate _destructSessionPvt;
+};
+
+static SessionPrivate SessionPvt;
+
+OpenCDMError OpenCDMSession::CreateSession(struct OpenCDMSystem* system,
+                            const LicenseType licenseType, const char initDataType[],
+                            const uint8_t initData[], const uint16_t initDataLength,
+                            const uint8_t CDMData[], const uint16_t CDMDataLength,
+                            OpenCDMSessionCallbacks* callbacks, void* userData,
+                            struct OpenCDMSession** session)
+{
+    OpenCDMError result(ERROR_INVALID_ACCESSOR);
+
+    if (system != nullptr) {
+        *session = new OpenCDMSession(system, std::string(initDataType),
+                            initData, initDataLength, CDMData,
+                            CDMDataLength, licenseType, callbacks, userData);
+        result = (*session != nullptr ? OpenCDMError::ERROR_NONE
+                                      : OpenCDMError::ERROR_INVALID_SESSION);
+        if(result == OpenCDMError::ERROR_NONE) {
+            void *pvtData = nullptr;
+            (*session)->_pvtData = nullptr;
+            OpenCDMError pvt_result(OpenCDMError::ERROR_METHOD_NOT_IMPLEMENTED);
+            pvt_result = SessionPvt.Construct(*session, pvtData);
```

Comment:
Is there a reason for not directly pushing the (*session)->_pvtData ? You could just call: 
pvt_result = SessionPvt.Construct(*session,  (*session)->_pvtData); 
Than you could also drop the following code: 
`
if(pvt_result == OpenCDMError::ERROR_NONE) {
    (*session)->_pvtData = pvtData;
}
else 
`

Reply:
Updated based on previous comment.